File: python-package/vbatten_x/physics.py
from __future__ import annotations

import json
from enum import Enum
from typing import Optional, Dict, Any


class PDEType(str, Enum):
    NONE          = "none"
    HEAT          = "heat"
    WAVE          = "wave"
    NAVIER_STOKES = "navier_stokes"
    POISSON       = "poisson"
    CUSTOM        = "custom"


class SymmetryGroup(str, Enum):
    ROTATION_2D  = "rotation_2d"
    ROTATION_3D  = "rotation_3d"
    REFLECTION   = "reflection"
    PERMUTATION  = "permutation"
    TRANSLATION  = "translation"


class BCType(str, Enum):
    DIRICHLET = "dirichlet"
    NEUMANN   = "neumann"
    PERIODIC  = "periodic"


class PhysicsSpec:
    def __init__(self):
        self._spec: Dict[str, Any] = {
            "pde_type":             PDEType.NONE.value,
            "pde_diffusivity":      1.0,
            "pde_dt":               0.01,
            "pde_viscosity":        1e-3,
            "grid_nx":              16,
            "grid_ny":              16,
            "spatial_resolution":   1.0,
            "symmetry_groups":      [],
            "conserved_quantities": [],
            "boundary_conditions":  {},
        }

    def pde(self,
            pde_type: PDEType,
            diffusivity: float = 1.0,
            dt: float = 0.01,
            viscosity: float = 1e-3) -> "PhysicsSpec":
        self._spec["pde_type"]        = pde_type.value if isinstance(pde_type, PDEType) else pde_type
        self._spec["pde_diffusivity"] = diffusivity
        self._spec["pde_dt"]          = dt
        self._spec["pde_viscosity"]   = viscosity
        return self

    def symmetry(self, group: SymmetryGroup) -> "PhysicsSpec":
        v = group.value if isinstance(group, SymmetryGroup) else group
        if v not in self._spec["symmetry_groups"]:
            self._spec["symmetry_groups"].append(v)
        return self

    def conserve(self, quantity: str) -> "PhysicsSpec":
        if quantity not in self._spec["conserved_quantities"]:
            self._spec["conserved_quantities"].append(quantity)
        return self

    def boundary(self,
                 region: str,
                 bc_type: BCType,
                 value: float = 0.0,
                 flux: float = 0.0) -> "PhysicsSpec":
        t = bc_type.value if isinstance(bc_type, BCType) else bc_type
        self._spec["boundary_conditions"][region] = {
            "type": t, "value": value, "flux": flux
        }
        return self

    def grid(self, nx: int, ny: int, resolution: float = 1.0) -> "PhysicsSpec":
        self._spec["grid_nx"]            = nx
        self._spec["grid_ny"]            = ny
        self._spec["spatial_resolution"] = resolution
        return self

    def to_json(self) -> str:
        return json.dumps(self._spec)

    @classmethod
    def from_json(cls, s: str) -> "PhysicsSpec":
        spec = cls()
        spec._spec = json.loads(s)
        return spec

    def __repr__(self) -> str:
        return f"PhysicsSpec(pde={self._spec['pde_type']}, symmetries={self._spec['symmetry_groups']})"

File: R-package/R/physics.R
# VBattenX Physics Spec builder for R

#' Create a PhysicsSpec builder
#' @return vbx.PhysicsSpec object
#' @export
vbx.physics.spec <- function() {
  spec <- list(
    pde_type             = "none",
    pde_diffusivity      = 1.0,
    pde_dt               = 0.01,
    pde_viscosity        = 1e-3,
    grid_nx              = 16L,
    grid_ny              = 16L,
    spatial_resolution   = 1.0,
    symmetry_groups      = list(),
    conserved_quantities = list(),
    boundary_conditions  = list()
  )
  structure(spec, class = "vbx.PhysicsSpec")
}

#' Set the PDE type on a PhysicsSpec
#' @export
vbx.pde <- function(spec, type, diffusivity = 1.0, dt = 0.01, viscosity = 1e-3) {
  stopifnot(inherits(spec, "vbx.PhysicsSpec"))
  spec$pde_type        <- type
  spec$pde_diffusivity <- diffusivity
  spec$pde_dt          <- dt
  spec$pde_viscosity   <- viscosity
  spec
}

#' Add a symmetry group
#' @export
vbx.symmetry <- function(spec, group) {
  stopifnot(inherits(spec, "vbx.PhysicsSpec"))
  if (!group %in% spec$symmetry_groups)
    spec$symmetry_groups <- c(spec$symmetry_groups, group)
  spec
}

#' Declare a conserved quantity
#' @export
vbx.conserve <- function(spec, quantity) {
  stopifnot(inherits(spec, "vbx.PhysicsSpec"))
  spec$conserved_quantities <- c(spec$conserved_quantities, quantity)
  spec
}

#' Set boundary condition
#' @export
vbx.boundary <- function(spec, region, type, value = 0.0, flux = 0.0) {
  stopifnot(inherits(spec, "vbx.PhysicsSpec"))
  spec$boundary_conditions[[region]] <- list(type = type, value = value, flux = flux)
  spec
}

#' Set grid dimensions
#' @export
vbx.grid <- function(spec, nx, ny, resolution = 1.0) {
  stopifnot(inherits(spec, "vbx.PhysicsSpec"))
  spec$grid_nx             <- as.integer(nx)
  spec$grid_ny             <- as.integer(ny)
  spec$spatial_resolution  <- resolution
  spec
}

#' Convert PhysicsSpec to JSON string
#' @export
vbx.spec.to_json <- function(spec) {
  jsonlite::toJSON(unclass(spec), auto_unbox = TRUE)
}

#' @export
print.vbx.PhysicsSpec <- function(x, ...) {
  cat("VBattenX PhysicsSpec\n")
  cat("  PDE:       ", x$pde_type, "\n")
  cat("  Symmetries:", paste(x$symmetry_groups, collapse = ", "), "\n")
  cat("  Conserved: ", paste(x$conserved_quantities, collapse = ", "), "\n")
  invisible(x)
}
